test/functional/runner/index.js

```diff
@@ -0,0 +1,590 @@
+    })
+    .then(() => {
+      if (suite.key_vault_data) {
+        const dataKeysCollection = context.sharedClient.db('admin').collection('datakeys');
+        return dataKeysCollection
+          .drop({ w: 'majority' })
+          .catch(err => {
+            if (!err.message.match(/ns not found/)) {
+              throw err;
+            }
+          })
+          .then(() => {
+            if (suite.key_vault_data.length) {
+              return dataKeysCollection.insertMany(suite.key_vault_data, { w: 'majority' });
+            }
+          });
+      }
+    })
+    .then(() => {
+      const options = { w: 'majority' };
+      if (suite.json_schema) {
+        options.validator = { $jsonSchema: suite.json_schema };
+      }
+
+      return db.createCollection(context.collectionName, options);
+    })
+    .then(() => {
+      if (suite.data && Array.isArray(suite.data) && suite.data.length > 0) {
+        return coll.insertMany(suite.data, { w: 'majority' });
+      }
+    })
+    .then(() => {
+      return context.runForAllClients(client => {
+        return client
+          .db(context.dbName)
+          .collection(context.collectionName)
+          .distinct('x')
+          .catch(() => {});
+      });
+    });
+}
+
+function parseSessionOptions(options) {
+  const result = Object.assign({}, options);
+  if (result.defaultTransactionOptions && result.defaultTransactionOptions.readPreference) {
+    result.defaultTransactionOptions.readPreference = normalizeReadPreference(
+      result.defaultTransactionOptions.readPreference.mode
+    );
+  }
+
+  return result;
+}
+
+const IGNORED_COMMANDS = new Set(['ismaster']);
+
+let displayCommands = false;
+function runTestSuiteTest(configuration, spec, context) {
+  const commandEvents = [];
+  const clientOptions = translateClientOptions(
+    Object.assign({ monitorCommands: true }, spec.clientOptions)
+  );
+
+  // test-specific client options
+  clientOptions.autoReconnect = false;
+  clientOptions.haInterval = 100;
+  clientOptions.useRecoveryToken = true;
+
+  // TODO: this should be configured by `newClient` and env variables
+  clientOptions.useUnifiedTopology = true;
+
+  const url = resolveConnectionString(configuration, spec);
+  const client = configuration.newClient(url, clientOptions);
+  return client.connect().then(client => {
+    context.testClient = client;
+    client.on('commandStarted', event => {
+      if (IGNORED_COMMANDS.has(event.commandName)) {
+        return;
+      }
+
+      // if (event.databaseName === context.dbName || isTransactionCommand(event.commandName)) {
+      commandEvents.push(event);
+
+      // very useful for debugging
+      if (displayCommands) {
+        console.dir(event, { depth: 5 });
+      }
+    });
+
+    const sessionOptions = Object.assign({}, spec.transactionOptions);
+
+    spec.sessionOptions = spec.sessionOptions || {};
+    const database = client.db(context.dbName);
+    const session0 = client.startSession(
```

Comment:
are `session0` and `session1` universal?

Reply:
yes, they are defined as part of the spec runner README.md